Sources/Yams/Encoder.swift

```diff
@@ -0,0 +1,344 @@
+
+        func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type) -> KeyedEncodingContainer<NestedKey> {
+            return currentEncoder.container(keyedBy: type)
+        }
+
+        func nestedUnkeyedContainer() -> UnkeyedEncodingContainer { return currentEncoder.unkeyedContainer() }
+        func superEncoder() -> Encoder { return currentEncoder }
+
+        // MARK: -
+
+        private var currentEncoder: _YAMLReferencingEncoder {
+            defer { encoder.sequence.append("") }
+            return encoder.encoder(at: count)
+        }
+    }
+
+    extension _YAMLEncoder: SingleValueEncodingContainer {
+
+        // MARK: - Swift.SingleValueEncodingContainer Methods
+
+        func encodeNil() throws {
+            assertCanEncodeNewValue()
+            node = .null
+        }
+
+        func encode(_ value: Bool)   throws { try represent(value) }
+        func encode(_ value: Int)    throws { try represent(value) }
+        func encode(_ value: Int8)   throws { try represent(value) }
+        func encode(_ value: Int16)  throws { try represent(value) }
+        func encode(_ value: Int32)  throws { try represent(value) }
+        func encode(_ value: Int64)  throws { try represent(value) }
+        func encode(_ value: UInt)   throws { try represent(value) }
+        func encode(_ value: UInt8)  throws { try represent(value) }
+        func encode(_ value: UInt16) throws { try represent(value) }
+        func encode(_ value: UInt32) throws { try represent(value) }
+        func encode(_ value: UInt64) throws { try represent(value) }
+        func encode(_ value: Float)  throws { try represent(value) }
+        func encode(_ value: Double) throws { try represent(value) }
+
+        func encode(_ value: String) throws {
+            assertCanEncodeNewValue()
+            node = Node(value)
+        }
+
+        func encode<T>(_ value: T) throws where T: Encodable {
+            assertCanEncodeNewValue()
+            if let customized = value as? ScalarRepresentableCustomizedForCodable {
+                node = customized.representedForCodable()
+            } else if let representable = value as? ScalarRepresentable {
+                node = try box(representable)
+            } else {
+                try value.encode(to: self)
+            }
+        }
+
+        // MARK: -
+
+        /// Asserts that a single value can be encoded at the current coding path
+        /// (i.e. that one has not already been encoded through this container).
+        /// `preconditionFailure()`s if one cannot be encoded.
+        private func assertCanEncodeNewValue() {
+            precondition(
+                canEncodeNewValue,
+                "Attempt to encode value through single value container when previously value already encoded."
+            )
+        }
+    }
+
+    // MARK: - CodingKey for `_UnkeyedEncodingContainer` and `superEncoders`
+
+    struct _YAMLEncodingKey: CodingKey { // swiftlint:disable:this type_name
```

Comment:
This shares an implementation with `_YAMLDecodingKey`. Why not merge them into a `_YAMLCodingKey`?